BLE_GattServer_FOTAService/MCUboot/client/client.py

```diff
@@ -0,0 +1,330 @@
+                if count >= 5:
+                    raise asyncio.TimeoutError
+
+    async def transfer_binary(self):
+        binary_sent = False
+
+        while not binary_sent:
+            self.check_status_notifications()
+
+            if self.flow_paused:
+                continue
+
+            n = (MAXIMUM_FRAGMENT_ID + 1) * self.rollover_counter + self.fragment_id
+            packet_data = bytearray(get_chunk_n(self.data, FRAGMENT_SIZE, n))
+            if len(packet_data) < FRAGMENT_SIZE:
+                binary_sent = True
+                if len(packet_data) == 0:
+                    continue
+            fragment = bytearray([self.fragment_id]) + packet_data
+            try:
+                await asyncio.wait_for(self.client.write_gatt_char(UUID_BINARY_STREAM_CHAR, fragment, False),
+                                       timeout=0.025)
+                await asyncio.sleep(0.075)
+            except asyncio.TimeoutError:
+                logger.error(f'Timeout error occurred while writing BSC')
+                await asyncio.sleep(0.100)
+
+            self.fragment_id += 1
+            # In case of rollover, increment counter and reset fragment ID
+            if  self.fragment_id >= MAXIMUM_FRAGMENT_ID + 1:
+                self.rollover_counter += 1
+                self.fragment_id = 0
+
+        await self.client.write_gatt_char(UUID_CONTROL_CHAR, FOTA_OP_CODE_COMMIT, True)
+
+
+async def main():
+    client = await find_device('FOTA')
+
+    if client is None:
+        return
+
+    def disconnected_callback(c: BleakClient):
+        disconnection_complete_event.set()
+
+    async def waiter(event: asyncio.Event):
+        await event.wait()
+
+    client.set_disconnected_callback(disconnected_callback)
+    disconnection_complete_event = asyncio.Event()
+    disconnection_complete = asyncio.create_task(waiter(disconnection_complete_event))
+
+    path = input('Enter the path to the binary: ')
+
+    try:
+        file = open(path, 'rb')
+    except IOError:
+        logger.error(f'Could not open file {path}')
+        await client.disconnect()
+        raise
+    data = file.read()
+
+    session = FOTASession(client, data)
+
+    fw_rev, dev_str = await session.get_firmware_revision()
+    logger.info(f'DFU Service found with firmware rev {fw_rev.decode("utf-8")}' +
+               (f' for device "{dev_str.decode("utf-8")}"' if dev_str else ''))
+
+    try:
+        await session.start()
+    except asyncio.TimeoutError:
+        logger.error(f'FOTA session failed to start within timeout period')
+        await client.disconnect()
+        return
+
+    logger.info(f'FOTA session started')
+
+    logger.info(f'Transferring binary of size {len(data)} bytes, '
+                f'hash: {hashlib.sha256(data).hexdigest()}')
+
+    await session.transfer_binary()
+
+    await disconnection_complete
+
+    logger.info("FOTA session complete, waiting for device to apply update")
+    await asyncio.sleep(60.0)
```

Comment:
Would it be possible to actively search for the device for 60 seconds ?

Reply:
It probably makes more sense to search for the device until we are connected.